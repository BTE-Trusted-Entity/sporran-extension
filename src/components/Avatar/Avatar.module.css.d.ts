export const __esModule: true;
export const avatar: string;
export const identicon: string;
export const fullDid: string;
