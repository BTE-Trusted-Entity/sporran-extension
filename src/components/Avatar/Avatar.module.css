.avatar {
  width: calc(84rem / 16);
  height: calc(84rem / 16);
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  background: rgb(var(--color-white-rgb) / 30%) no-repeat;
}

.identicon {
  --size: calc(72% / 84 * 100);

  width: var(--size);
  height: var(--size);
  border-radius: 50%;
}

.fullDid {
  background: rgb(var(--color-orange-rgb) / 60%) no-repeat;
}
