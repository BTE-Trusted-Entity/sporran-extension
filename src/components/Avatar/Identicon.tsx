import React from 'react';
import { polkadotIcon } from '@polkadot/ui-shared';

interface Props {
  address: string;
  className?: string;
  size: number;
}

export function Identicon({
  address,
  className = '',
  size,
}: Props): JSX.Element {
  const circles = polkadotIcon(address, { isAlternative: false });
  return (
    <svg className={className} height={size} width={size} viewBox="0 0 64 64">
      {circles.map(({ cx, cy, fill, r }, key) => (
        <circle cx={cx} cy={cy} fill={fill} r={r} key={key} />
      ))}
    </svg>
  );
}
