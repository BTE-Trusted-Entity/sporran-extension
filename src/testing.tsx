export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
