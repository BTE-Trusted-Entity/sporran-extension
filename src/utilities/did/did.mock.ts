import { isFullDid } from './did';

jest.mock('./did');

export function mockIsFullDid(boolean: boolean): void {
  jest.mocked(isFullDid).mockReturnValue(boolean);
}
