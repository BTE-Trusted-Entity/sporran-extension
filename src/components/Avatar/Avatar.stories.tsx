import { Meta } from '@storybook/react';

import { identitiesMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { Avatar } from './Avatar';

export default {
  title: 'Components/Avatar',
  component: Avatar,
  decorators: [(story) => <div style={{ textAlign: 'center' }}>{story()}</div>],
} as Meta;

export function Template(): JSX.Element {
  return (
    <Avatar
      identity={
        identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
      }
    />
  );
}

export function withFullDid(): JSX.Element {
  return (
    <Avatar
      identity={
        identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
      }
    />
  );
}
