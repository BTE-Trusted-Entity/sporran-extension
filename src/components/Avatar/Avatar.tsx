import cx from 'classnames';

import * as styles from './Avatar.module.css';

import { isFullDid } from '../../utilities/did/did';
import { Identity } from '../../utilities/identities/types';

import { Identicon } from './Identicon';

interface Props {
  identity: Identity;
  className?: string;
}

export function Avatar({
  identity,
  className = styles.avatar,
}: Props): JSX.Element {
  return (
    <div
      className={cx(className, {
        [styles.fullDid]: isFullDid(identity.did),
      })}
    >
      <Identicon
        className={styles.identicon}
        address={identity.address}
        size={64}
      />
    </div>
  );
}
