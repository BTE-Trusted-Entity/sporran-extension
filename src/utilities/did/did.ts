import { resolve, Utils } from '@kiltprotocol/did';
import { DidDocument, DidEncryptionKey, DidUri } from '@kiltprotocol/types';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Utils.parseDidUri(did).type === 'full';
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  const details = await resolve(did);
  if (!details?.document) {
    throw new Error(`Cannot resolve DID ${did}`);
  }

  return details.document;
}

export function useDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getDidDocument, [did]);
}

export function parseDidUri(did: DidUri): ReturnType<
  typeof Utils.parseDidUri
> & {
  lightDid: DidUri;
  fullDid: DidUri;
} {
  const parsed = Utils.parseDidUri(did);
  const { address, type } = parsed;

  const lightDid =
    type === 'light' ? did : (`did:kilt:light:00${address}` as DidUri);

  const fullDid = type === 'full' ? did : Utils.getFullDidUri(did);

  return {
    ...parsed,
    lightDid,
    fullDid,
  };
}

export function getDidEncryptionKey(document: DidDocument): DidEncryptionKey {
  const { keyAgreement } = document;
  if (!keyAgreement?.length) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}
