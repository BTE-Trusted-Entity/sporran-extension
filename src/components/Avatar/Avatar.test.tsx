import { identitiesMock, render } from '../../testing/testing';

import { mockIsFullDid } from '../../utilities/did/did.mock';

import { Avatar } from './Avatar';

describe('Avatar', () => {
  it('should render', async () => {
    const { container } = render(
      <Avatar
        identity={
          identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
        }
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render full DID avatar style', async () => {
    mockIsFullDid(true);
    const { container } = render(
      <Avatar
        identity={
          identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
