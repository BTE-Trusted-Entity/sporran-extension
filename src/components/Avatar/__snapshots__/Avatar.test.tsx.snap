// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render 1`] = `
<div>
  <div
    class="avatar"
  >
    Identicon
  </div>
</div>
`;

exports[`Avatar should render full DID avatar style 1`] = `
<div>
  <div
    class="avatar fullDid"
  >
    Identicon
  </div>
</div>
`;
